### PR TITLE
Add conversion for tosa.transpose

### PR DESCRIPTION
**Pull request by OliverScherf**

### CODE REVIEW

File: include/emitc/emitc_tosa.h [hunk 1]
```diff
@@ -351,6 +351,35 @@ inline Dest reshape(Src x) {
...
 
+// TransposeOp
+// Maps the perms dimension from Dest to Src
+template <typename Dest, typename Src>
+inline Dest transpose(Src operand, Tensor1D<int64_t, Src::rank()> perms) {
+  static_assert(is_tensor<Src>::value, "Expected tensor argument");
+  static_assert(is_tensor<Dest>::value, "Expected tensor result");
+
+  // Since broadcast_dimensions map the broadcast_dimensions from Src to Dest,
+  // we have to reverse the tosa.transpose perms mapping rom Dest to Src
```
Comment (dajuro) [inline include/emitc/emitc_tosa.h:362]:
```suggestion
  // we have to reverse the tosa.transpose perms mapping from Dest to Src
```
typo

**Review by simon-camp — CHANGES_REQUESTED** [review]
LGTM besides the nits.

File: unittests/emitc_tosa.cpp [hunk 2]
```diff
@@ -266,4 +266,57 @@ TEST(tosa, reshape) {
...
+TEST(tosa, transpose) {
+  // clang-format off
+  Tensor2D<float, 3, 2> t0 = {1, 2,
+                              3, 4,
+                              5, 6};
+  Tensor1D<int32_t, 2> perms_i32 = {1, 0};
+  Tensor1D<int64_t, 2> perms_i64 = {1, 0};
+  Tensor1D<int64_t, 2> no_perms =  {0, 1};
+  Tensor2D<float, 2, 3> expected_result0 = {1, 3, 5,
+
```
Comment (dajuro) [inline unittests/emitc_tosa.cpp:278]:
```suggestion

```
Is this empty line on purpose?

**Review by dajuro — CHANGES_REQUESTED** [review]
LGTM some minor tweaks required...

File: include/emitc/emitc_tosa.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (dajuro) [reply include/emitc/emitc_tosa.h:362]: This is easier to read :+1:

[the remark this answers is not in the document]
Comment (marbre) [reply include/emitc/emitc_tosa.h:362]: Your suggestion changes broadcast_dimensions to broadcast_in_dim. Is this intended? But I agree, this one is easier.

File: unittests/emitc_tosa.cpp (hunk 2 above)
Comment (OliverScherf) [reply unittests/emitc_tosa.cpp]: No, that happens when you turn of clang-format :)

File: include/emitc/emitc_tosa.h (hunk 1 above)
Comment (OliverScherf) [reply include/emitc/emitc_tosa.h]: You're right, but the comment was updated anyway

[the remark this answers is not in the document]
Comment (OliverScherf) [reply include/emitc/emitc_tosa.h]: I tweaked the comment further, it should be more clear now.

File: include/emitc/emitc_tosa.h [hunk 3]
```diff
@@ -351,6 +351,35 @@ inline Dest reshape(Src x) {
...
+  }
+  return emitc::broadcast_in_dim<Dest>(operand, broadcast_dimensions);
+}
+
+// TransposeOp allows perms to be of type int32_t or int64_t.
+template <typename Dest, typename Src>
+inline Dest transpose(Src input, Tensor1D<int32_t, Src::rank()> perms) {
+  Tensor1D<int64_t, Src::rank()> permsInt64;
+  for (size_t i = 0; i < perms.size(); ++i) {
+    permsInt64[i] = perms[i];
```
[the remark this answers is not in the document]
Comment (OliverScherf) [reply include/emitc/emitc_tosa.h]: added :+1:

File: include/emitc/emitc_mhlo.h [hunk 4]
```diff
@@ -328,6 +328,7 @@ inline Src logical_xor(Src x, Src y) {
 
 /// Functions for other MHLO ops.
 // BroadcastInDimOp
+// Maps the broadcast_dimensions from Src to Dest
```
Comment (marbre) [inline include/emitc/emitc_mhlo.h:331]:
```suggestion
// Maps the broadcast dimensions from Src to Dest
```
?

Comment (marbre) [reply include/emitc/emitc_mhlo.h]: Oh, even better with your latest commit :)

**Review by marbre — APPROVED** [review]

File: include/emitc/emitc_mhlo.h (hunk 4 above)
Comment (OliverScherf) [reply include/emitc/emitc_mhlo.h]: Just updated the comment as you were typing your review! :)

File: unittests/emitc_tosa.cpp [hunk 5]
```diff
@@ -266,4 +266,56 @@ TEST(tosa, reshape) {
...
+                                               4, 8, 12, 16, 20, 24};
+  // clang-format on
+  Tensor3D<float, 4, 2, 3> s2 =
+      tosa::transpose<Tensor3D<float, 4, 2, 3>>(t2, perms2);
+  EXPECT_THAT(s2, Pointwise(Eq(), expected_result2));
+
+  Tensor<float, 13, 21, 3> v1;
+  Tensor<int32_t, 3> v2{2, 0, 1};
+  Tensor<float, 3, 13, 21> v3 =
+      tosa::transpose<Tensor<float, 3, 13, 21>>(v1, v2);
```
Comment (simon-camp) [inline unittests/emitc_tosa.cpp:318]: Is this unused? Or is it just to check that the statements compile?

Comment (dajuro) [reply unittests/emitc_tosa.cpp:318]: Yep, this looks suspicious!

Comment (OliverScherf) [reply unittests/emitc_tosa.cpp]: Indeed unused, I removed it

**Review by dajuro — APPROVED** [review]
LGTM :+1:

**Review by marbre — APPROVED** [review]

**Review by simon-camp — APPROVED** [review]

**Pull request merged**